ownCloud/Settings/Passcode/PasscodeManager.swift

```diff
@@ -0,0 +1,375 @@
+        }
+
+        super.init()
+    }
+
+    // MARK: - Show Passcode View
+
+    func showPasscodeIfNeeded(hiddenOverlay:Bool) {
+
+        if isPasscodeActivated {
+            if self.passcodeViewController == nil {
+
+                self.completionHandler = {
+                    self.datePressedHomeButton = nil
+                    self.timesPasscodeFailed = 0
+                    self.dismissPasscode()
+                }
+
+                self.passcodeViewController = PasscodeViewController(hiddenOverlay:hiddenOverlay)
+                self.createLockWindow()
+
+                // Brute force protection
+                if let date = self.dateAllowTryAgain, date > Date() {
+                    //User killed the app
+                    self.passcodeMode = .unlockPasscodeError
+                    self.passcodeViewController?.setEnableNumberButtons(isEnable: false)
+                    self.scheduledTimerToUpdateInterfaceTime()
+                } else {
+                    self.passcodeMode = .unlockPasscode
+                }
+
+                self.updateUI()
+
+            } else {
+                self.passcodeViewController?.showOverlay()
+            }
+        }
+    }
+
+    func showAddOrEditPasscode(completionHandler: @escaping CompletionHandler) {
+
+        if isPasscodeActivated {
+            self.passcodeMode = PasscodeInterfaceMode.deletePasscode
+        } else {
+            self.passcodeMode = PasscodeInterfaceMode.addPasscodeFirstStep
+        }
+
+        self.completionHandler = completionHandler
+
+        self.passcodeViewController = PasscodeViewController(hiddenOverlay:true)
+        self.createLockWindow()
+
+        self.updateUI()
+
+        self.window?.showWindowAnimation()
+    }
+
+    func createLockWindow() {
+        self.window = LockWindow(frame: UIScreen.main.bounds)
+        self.window?.windowLevel = UIWindowLevelStatusBar
+        self.window?.rootViewController = self.passcodeViewController!
+        self.window?.makeKeyAndVisible()
+    }
+
+    func homeButtonPressed() {
+
+        //Protection to hide the PasscodeViewController if is on Delete modes to show the Unlock
+        if self.passcodeViewController != nil {
+            if self.passcodeMode == .deletePasscode ||
+                self.passcodeMode == .deletePasscodeError {
+                self.completionHandler!()
+                self.window?.isHidden = true
+                self.passcodeViewController = nil
+            }
+        }
+
+        //Store the date when the user pressed home button
+        if self.isPasscodeActivated, self.datePressedHomeButton == nil {
+            self.datePressedHomeButton = Date()
+        }
+
+        //Show the passcode
+        self.showPasscodeIfNeeded(hiddenOverlay: false)
+    }
+
+    // MARK: - Interface updates
+
+    private func updateUI() {
+
+        var messageText : String?
+        var errorText : String? = ""
```

Comment:
`errorText` here is optional but you are assigning it value and as I see in this method never has nil value so consider make it non-optional.